### ISSUE TITLE
Catalog endpoint does not seem to be captured by full access?

### ISSUE THREAD

**Issue by schancel** [cesanta/docker_auth#201]
I have a service account with the following permissions:
```
  - match: {account: "ahab"}
    actions: ["*"]
    comment: "Captain-ahab has full access to everything."
```

And it is getting this error while querying the catalog:
```
{  
   "errors":[  
      {  
         "code":"UNAUTHORIZED",
         "message":"authentication required",
         "detail":[  
            {  
               "Type":"registry",
               "Class":"",
               "Name":"catalog",
               "Action":"*"
            }
         ]
      }
   ]
}
```

Is there something additional required for "full access to everything?"

**Comment by adamdecaf** [comment]
I've been trying to debug the same issue for `/v2/$name/tags/list`. @rojer Do you know by chance?

**Comment by adamdecaf** [comment]
I think my problem is the same as @schancel, but I'm not sure where to go next.

Here are my ACLs: 

```yaml
acl:
  - match: {account: "jenkins"}
    actions: ["pull", "push"]
    comment: "Jenkins can push and all images"
  - match: {type: "repository", name: "/.+/"}
    actions: ["pull"]
    comment: "Let anonymous accounts view repository information (i.e. image tags)"
  - match: {team: "infrastructure"}
    actions: ["pull","push"]
    comment: "Infrastructure team members can push and all images"
  - match: {account: ""}
    actions: ["pull","push"]
    comment: "Anonymous users can pull (and for now push) all images"
```

I get the following response when trying to view tags (`/v2/$name/tags/list`): 

```json
{
    "errors": [
        {
            "code": "UNAUTHORIZED",
            "message": "authentication required",
            "detail": [
                {
                    "Type": "repository",
                    "Class": "",
                    "Name": "elasticsearch-curator",
                    "Action": "pull"
                }
            ]
        }
    ]
}
```

docker-auth logs: 
```
I0626 20:15:44.719786       1 server.go:375] Auth request: {:@10.10.1.95:50184 [{repository alerts-settings-server [pull]}]}

I0626 20:15:44.719812       1 server.go:217] Authn static  -> true, map[], <nil>

I0626 20:15:44.719883       1 acl.go:116] { pull repository alerts-settings-server} matched {"Match":{"type":"repository","name":"/.+/"},"Actions":["pull"],"Comment":null} (Comment: %!s(*string=<nil>))

I0626 20:15:44.719960       1 server.go:239] Authz static ACL { pull repository alerts-settings-server} -> [pull], %!s(<nil>)

I0626 20:15:44.726126       1 server.go:329] New token for {:@10.10.1.95:50184 [{repository alerts-settings-server [pull]}]} map[]: {"iss":"banno-docker-auth","sub":"","aud":"registry.banno-internal.com","exp":...,"nbf":...,"iat":...,"jti":"...","access":[{"type":"repository","name":"alerts-settings-server","actions":["pull"]}]}
```

I do have a question in that why is `alerts-settings-server` the name here? I've only made requests to another name `elasticsearch-sidecar`.

**Comment by adamdecaf** [comment]
After a couple days struggling to figure this out I think I've got a fix. cc @schancel 

tl;dr You need to request a token and use that on your `/_catalog` or `/tag/list` request. 

https://github.com/docker/distribution/issues/877#issuecomment-131197824 mentions "all endpoints are protected" and after reading the docker registry code that's true. Thus, any request _without auth credentials_ will be rejected by the registry (and not even hit docker_auth). 

If you [request a token](https://docs.docker.com/registry/spec/auth/token/#requesting-a-token) and use that then I'm able to list tags while the registry is under auth. 

It's easy enough to emulate this with curl:

```
$ curl -v -H "Authorization: Bearer dGVzdAo=" https://registry.company.com/v2/elasticsearch-curator/tags/list 
...
www-authenticate: Bearer realm="https://registry.company.com:5001/auth",service="registry.company.com",scope="repository:elasticsearch-curator:pull",error="invalid_token"
```

Then, make the token request back at `docker_auth`: 

```
$ curl "https://registry.company.com:5001/auth?service=registry.company.com&scope=repository:elasticsearch-curator:pull,pull"
{"token":"..."}
```

Using that auth token works now:

```
$ curl -v -H "Authorization: Bearer ..." https://registry.company.com/v2/elasticsearch-curator/tags/list 
{"name":"elasticsearch-curator","tags":["latest"]}
```

**Comment by jomach** [comment]
I had the same problem when trying to query the endpoint /v2/_catalog via firefox. At the end I get a token on on auth endpoint without any authentication for the service that I want and then it works.

**Comment by bachka** [comment]
@jomach hi there, i am stuck now with this. what do you mean "get a token without any authentication on endpoint of auth" ? i tried all possibilites but somehow can not check why i am not getting /_catalog. it tells me still "invalid token". Can you help me further?
Thanks a lot!

**Comment by jomach** [comment]
@bachka maybe try 
`- match: {account: "*"}
    actions: ["pull","push"]
    comment: "Anonymous users can pull (and for now push) all images"`
 to allow Anonymous Users.

**Comment by dxlr8r** [comment]
I have the same issue with `_catalog`. Listing however works, my config:

```
acl:
  - match: {account: "admin"}
    actions: ["*"]
    comment: "admin: full access"
```

To list tags for test/busybox, first request a token, and define the service and scope.

```
$ TOKEN=$(curl -s -H 'Content-Type: application/json' -X POST -d '{"username": "'"$UNAME"'", "password": "'"$UPASS"'"}' 'https://'"$server"'/auth/token?service=auth-token-service&scope=repository:test/busybox:pull' | jq -r .token)
```

Then list, same above but with a different scope:

```
TOKEN=$(curl -s -H 'Content-Type: application/json' -X POST -d '{"username": "'"$UNAME"'", "password": "'"$UPASS"'"}' 'https://'"$server"'/auth/token?service=auth-token-service&scope=registry:catalog:*' | jq -r .token)
```

```
$ curl -H "Authorization: Bearer ${TOKEN}" -X GET https://$server/v2/_catalog
{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"registry","Class":"","Name":"catalog","Action":"*"}]}]}
```

Using `curl` with -v the Www-Authenticate also states:

`< Www-Authenticate: Bearer realm="https://spc-simenya-bksl-registry.services.basefarm.net/auth/token",service="auth-token-service",scope="registry:catalog:*",error="insufficient_scope"`

So it seems like the token I got does not have access to that scope, even though I defined that I wanted access to that scope.


```
$ curl -H "Authorization: Bearer ${TOKEN}" -X GET https://$server/v2/test/busybox/tags/list
{"name":"test/busybox","tags":["latest"]}
```

So I can verify that authentication works.

For catalog, it does not however:

**Comment by dxlr8r** [comment]
I played around with this today as well. Using simple config:

```
  - match: {account: "admin"}
    actions: ["*"]
    comment: "admin: full access"
```

No need for `{account: "admin", type: "registry", name: "catalog"}` etc. in this instance.

I believe many forget to set the scope, two examples, in both examples remember to set `id`, `pw` and `server`:

----

Set token:

```
token=$(curl -sH 'Content-Type: application/json' --user $id:$pw 'https://'"$server"'/auth/token?service=auth-token-service&scope=repository:test/busybox:pull' | jq -r '.token')
```

List tags for `test/busybox`:

```
curl -H "Authorization: Bearer ${token}" -X GET https://$server/v2/test/busybox/tags/list
```

----

And same as above, except change scope:

```
token=$(curl -sH 'Content-Type: application/json' --user $id:$pw 'https://'"$server"'/auth/token?service=auth-token-service&scope=registry:catalog:*' | jq -r '.token')
```

List catalog:

```
curl -H "Authorization: Bearer ${token}" -X GET https://$server/v2/_catalog
```

**Comment by techknowlogick** [comment]
Thank you @dxlr8r for your comment above. I was able to replicate what you have and confirmed this issues appears it can be closed.
